src/frontend/lib/MessagePortLike.ts

```diff
@@ -0,0 +1,119 @@
+import {EventSubscription} from 'react-native';
+import EventEmitter from 'react-native/Libraries/vendor/emitter/EventEmitter.js';
+import {Channel} from 'nodejs-mobile-react-native';
+
+type Listener = (...args: any[]) => void;
+
+export type ServerState = 'idle' | 'started' | 'closed';
+
+export class MessagePortLike extends EventEmitter {
+  #API_EVENT_NAME = '@@API_MESSAGE';
+  #channel: Channel;
+  #eventsSubscriptions = new Map<
+    string,
+    WeakMap<Listener, EventSubscription>
+  >();
+  #channelSubscription: EventSubscription;
+  #state: ServerState = 'idle';
+  #queuedMessages: any[] = [];
+
+  constructor(channel: Channel) {
+    super();
+    this.#channel = channel;
+
+    const handleChannelMessage = (message: any) => {
+      if (this.#state === 'idle') {
+        this.#queuedMessages.push(message);
+      } else if (this.#state === 'started') {
+        this.emit('message', message);
+      } else {
+        // no-op if the port is closed
+        // (the event listener should be removed anyway)
+      }
+    };
+
+    // @ts-expect-error
+    this.#channelSubscription = channel.addListener(
+      this.#API_EVENT_NAME,
+      handleChannelMessage,
+    );
+
+    channel.addListener('message', ({value}) => {
+      if (value === 'started') {
+        console.log('server started');
+      }
+    });
+  }
+
+  postMessage(message: any) {
+    this.#channel.post(this.#API_EVENT_NAME, message);
+  }
+
+  start() {
+    if (this.#state !== 'idle') {
+      return;
+    }
+    this.#state = 'started';
+    for (const msg of this.#queuedMessages) {
+      this.emit('message', msg);
+    }
+    this.#queuedMessages = [];
+  }
+
+  on(event: string, listener: (...args: any[]) => any) {
+    const sub = this.addListener(event, listener);
+
+    const registry = this.#eventsSubscriptions.get(event);
+
+    if (!registry) {
+      this.#eventsSubscriptions.set(event, new WeakMap([[listener, sub]]));
+      return this;
+    }
+
+    if (!registry.has(listener)) {
+      registry.set(listener, sub);
+    }
+
+    return this;
+  }
+
+  off(event: string, listener: (...args: any[]) => void) {
+    return this.removeListener(event, listener);
+  }
+
+  removeListener(event: string, listener: (...args: any[]) => void) {
+    const registry = this.#eventsSubscriptions.get(event);
+
+    if (!registry) {
+      return;
+    }
+
+    const subscription = registry.get(listener);
+
+    if (subscription) {
+      subscription.remove();
+      registry.delete(listener);
+
+      // TODO: Call this.#subscriptions.delete(event) if no more listeners?
+    }
+
+    return this;
+  }
+
+  removeAllListeners() {
+    this.#eventsSubscriptions.clear();
+    this.removeAllListeners();
+  }
```

Comment:
This seems to be re-implementing an event emitter, which is a hard thing to do and is full of edge cases. Why not use an existing library and extend from that rather than the react-native eventemitter?

Reply:
think the context for this (as well as the other review comments) mostly lives with stuff that I updated via https://github.com/digidem/react-native-with-mapeo-core-next/pull/8 (see PR description)

but yeah if there's something we can use to simplify this, probably a better approach

Reply:
think the main trickiness was that since this is tied to nodejs mobile's Channel implementation, it doesn't have the expected interface as the traditional Node EventEmitter, which RPC reflector relies on. 

React Native's EventEmitter uses a subscription based API i.e. `addListener` returns a `Subscription` and there are no methods like `on` and `off`, hence why this class is reimplementing some of those while also tying it with the Nodejs mobile Channel

Reply:
Yes, but why extend MessagePortLike from the react-native EventEmitter? Why not extend EventEmitter from [events](https://www.npmjs.com/package/events) or [EventEmitter3](https://www.npmjs.com/package/eventemitter3) - those have the methods that rpc-reflector checks. I would have thought that would work fine? There is no relationship between the type of EventEmitter that Channel extends and what MessagePortLike needs to be, except how you internally handle attaching and removing listeners from the channel.

Reply:
ah right. didn't think about that when initially implementing but yeah that makes sense 👍

Reply:
yeah, because what you're doing here is essentially reimplementing an EventEmitter, and also creating a compatibility layer with the react native EventEmitter, which is a maintenance overhead, and probably needs a lot more work to catch all the edge cases that the npm libs already check for.

Reply:
right, so the actionable changes here would be:

1. use one of the aforementioned libraries to extend for this class, instead of React Native's EventEmitter

2. hook into the `on` and `off` methods in order to you manage the nodejs channel listeners as well, which probably still requires some tracking similar to what we have in the current implementation

Reply:
Don't think so because it relies on Node's EventEmitter, which isn't available in React Native's runtime

Reply:
`tiny-typed-emitter` is just [require('events').EventEmitter](https://github.com/binier/tiny-typed-emitter/blob/master/lib/index.js#L3). It works in node, but you need to have a polyfill for react native. You can just use [`events`](https://www.npmjs.com/package/events) which is handy because it has the same name - you just need to install it from npm in the frontend - or you can use EventEmitter3 with tiny-typed-emitter if you configure Metro to package `events` -> `EventEmitter3`